Source/WebKit/Platform/IPC/TransferRegion.h

```diff
@@ -0,0 +1,221 @@
+        return TransferBuffer { buffer.releaseNonNull() };
+    }
+    TransferBuffer(TransferBuffer&&) = default;
+    TransferBuffer& operator=(TransferBuffer&&) = default;
+
+    TransferBufferIdentifier identifier() const
+    {
+        return m_identifier;
+    }
+
+    std::optional<WebKit::SharedMemory::Handle> passHandle()
+    {
+        if (m_hasBeenPassed || !m_buffer)
+            return std::nullopt;
+        m_hasBeenPassed = true;
+        return m_buffer->createHandle(WebKit::SharedMemory::Protection::ReadWrite);
+    }
+
+    std::span<uint8_t> data()
+    {
+        if (!m_buffer)
+            return { };
+        return std::span<uint8_t> { reinterpret_cast<uint8_t*>(m_buffer->data()), m_buffer->size() };
+    }
+
+    size_t size() const
+    {
+        if (!m_buffer)
+            return 0;
+        return m_buffer->size();
+    }
+
+    bool hasBeenPassed() const { return m_hasBeenPassed; }
+private:
+    TransferBuffer() = default;
+    TransferBuffer(Ref<WebKit::SharedMemory> buffer)
+        : m_buffer { WTFMove(buffer) }
+    {
+    }
+
+    TransferBufferIdentifier m_identifier { TransferBufferIdentifier::generateThreadSafe() };
+    RefPtr<WebKit::SharedMemory> m_buffer;
+    bool m_hasBeenPassed { false };
+};
+
+
+class ScopedTransferRegionMapping {
+public:
+    ScopedTransferRegionMapping(Ref<WebKit::SharedMemory> memory, size_t size)
+        : m_memory { WTFMove(memory) }
+        , m_size { size }
+    {
+    }
+    ScopedTransferRegionMapping() = default;
+
+    std::span<uint8_t> data() const
+    {
+        if (!m_memory)
+            return { };
+        return std::span<uint8_t> { reinterpret_cast<uint8_t*>(m_memory->data()), m_size };
+    }
+private:
+    RefPtr<WebKit::SharedMemory> m_memory;
+    size_t m_size { 0 };
+};
+
+// Class to hold shared memory instances that have been sent to a particular receiver process.
+class TransferRegionPool : public ThreadSafeRefCounted<TransferRegionPool> {
+    WTF_MAKE_NONCOPYABLE(TransferRegionPool);
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    explicit TransferRegionPool(Connection&);
+    std::optional<ScopedTransferRegion> reserveRegion(size_t);
+    void invalidate();
+    void releaseUnusedMemory();
+
+private:
+    void releaseTransferBuffer(TransferBuffer&&);
+    void scheduleReleaseUnusedMemory() WTF_REQUIRES_LOCK(m_lock);
+    void unscheduleReleaseUnusedMemory() WTF_REQUIRES_LOCK(m_lock);
+
+    Lock m_lock;
+    Connection* m_connection WTF_GUARDED_BY_LOCK(m_lock);
+    RefPtr<RunLoop::DispatchTimer> m_releaseUnusedTimer WTF_GUARDED_BY_LOCK(m_lock);
+    std::optional<TransferBuffer> m_free WTF_GUARDED_BY_LOCK(m_lock);
+    friend class ScopedTransferRegion;
+};
+
+// Class to hold shared memory mappings from a particular sender process.
+class TransferRegionMapper : public ThreadSafeRefCounted<TransferRegionMapper> {
+    WTF_MAKE_NONCOPYABLE(TransferRegionMapper);
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    TransferRegionMapper() = default;
```

Comment:
Classes that subclass RefCounted/ThreadSafeRefCounted should have a private constructor and a public create() function.